src/main/protobuf/dwrf_proto.proto

```diff
@@ -159,13 +172,46 @@ message StripeInformation {
     optional uint64 footerLength = 4;
     optional uint64 numberOfRows = 5;
     optional uint64 rawDataSize = 6;
+    // Arbitrary metadata similar for stripe level encryption
+    // Optimization?? only set for run start, and reuse until next run
+    repeated bytes keyMetadata = 9;
 }
 
 message UserMetadataItem {
     required string name = 1;
     required bytes value = 2;
 }
 
+message EncryptionGroup {
+    // Sub trees in the schema that encryption should be applied
+    // Sub tree is identified by its root id
+    // describe constraints
+    repeated uint32 nodes = 1;
+
+    // Arbitrary binary representing key metadata. It could be identifier
+    // of key in KMS, encrypted DEK or other form of user defined key metadata.
+    // Optimization?? this is optional and can use first keyMetadata from stripe?
```

Comment:
@zzhao0 can you clarify the status of this optimization (is it a part of the file format? what are the implications)

Reply:
if keyMetadata doesn't exist, that implied the entire file share the same key. Hence get key from the first stripe (key for first stripe must be set, since it's always start of the run)

if key metadata exists. That means it's a file with footer encrypted using it's own key (likely also means it's a concatenated file, with footer re-written). We then can use the key recorded in the footer as is